hash.go

```diff
@@ -0,0 +1,99 @@
+package bn256
+
+// HashG1 implements a hashing function into the G1 group. It uses Fouque-Tibouchi encoding described here
+// https://tools.ietf.org/pdf/draft-irtf-cfrg-hash-to-curve-04.pdf
+func HashG1(msg []byte) *G1 {
+	return mapToCurve(hashToBase(msg))
+}
+
+func mapToCurve(t *gfP) *G1 {
+	one := *newGFp(1)
+
+	// calculate w = (s * t)/(1 + B + t^2)
+	// we calculate w0 = s * t * (1 + B + t^2) and inverse of it, so that w = (st)^2/w0
+	// and then later x3 = 1 + (1 + B + t^2)^4/w0^2
+	w := &gfP{}
+
+	// a = (1 + B + t^2)
+	a := &gfP{}
+	t2 := &gfP{}
+	gfpMul(t2, t, t)
+	gfpAdd(a, curveB, t2)
+	gfpAdd(a, a, &one)
+
+	st := &gfP{}
+	gfpMul(st, s, t)
+
+	w0 := &gfP{}
+	gfpMul(w0, st, a)
+	w0.Invert(w0)
+
+	gfpMul(w, st, st)
+	gfpMul(w, w, w0)
+
+	e := legendre(t)
```

Comment:
`e` should not be Legendre (as in FT paper), in the draft this was replaced by `e = sign0(t)`

Reply:
done.